### PR TITLE
feat(scanner): Tag-based scan

### PR DESCRIPTION
**Pull request by MrCloudSec**
### Description

Add tags based scan using `-t/--scan-tags` option with a list of tags (Key=Value Key=Value ...).

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: mkdocs.yml [hunk 1]
```diff
@@ -41,6 +41,7 @@ nav:
           - Assume Role: tutorials/aws/role-assumption.md
           - AWS Security Hub: tutorials/aws/securityhub.md
           - AWS Organizations: tutorials/aws/organizations.md
+          - Tag-based Scan: tutorials/aws/tag-based-scan.md
           - Scan Multiple AWS Accounts: tutorials/aws/multiaccount.md
           - AWS CloudShell: tutorials/aws/cloudshell.md
           - Checks v2 to v3 Mapping: tutorials/aws/v2_to_v3_checks_mapping.md
```
Comment (toniblyx) [inline mkdocs.yml:46]:
```suggestion
          - Checks v2 to v3 Mapping: tutorials/aws/v2_to_v3_checks_mapping.md
          - Tag-based Scan: tutorials/aws/tag-based-scan.md
```

File: mkdocs.yml [hunk 2]
```diff
@@ -41,6 +41,7 @@ nav:
           - Assume Role: tutorials/aws/role-assumption.md
           - AWS Security Hub: tutorials/aws/securityhub.md
           - AWS Organizations: tutorials/aws/organizations.md
+          - Tag-based Scan: tutorials/aws/tag-based-scan.md
```
Comment (toniblyx) [inline mkdocs.yml]:
```suggestion
```

**Review by toniblyx — DISMISSED** [review]

File: prowler/lib/cli/parser.py [hunk 3]
```diff
@@ -343,6 +343,15 @@ def __init_aws_parser__(self):
...
         )
+        # Allowlist
+        audit_tags_subparser = aws_parser.add_argument_group("Tags-based Scan")
+        audit_tags_subparser.add_argument(
+            "-t",
+            "--scan-tags",
+            nargs="+",
+            default=None,
+            help="Scan only resources with specific tags (Key=Value), e.g., Environment=dev Project=prowler",
+        )
```
Comment (jfagoagas) [inline prowler/lib/cli/parser.py:354]: Please, include the parser test for these new arguments in `tests/lib/cli/parser_test.py`

File: prowler/lib/scan_filters/scan_filters.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
+from prowler.lib.logger import logger
+
+
+def is_resource_filtered(resource: str, audit_resources: list) -> bool:
+    """
+    Check if it is a tags-based
```
Comment (jfagoagas) [inline prowler/lib/scan_filters/scan_filters.py]:
```suggestion
    Check if the resource passed as argument is present in the audit_resources.
```

File: prowler/lib/scan_filters/scan_filters.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from prowler.lib.logger import logger
+
+
+def is_resource_filtered(resource: str, audit_resources: list) -> bool:
```
Comment (jfagoagas) [inline prowler/lib/scan_filters/scan_filters.py:4]:
```suggestion
def is_resource_filtered(resource: str, filtered_audit_resources: list) -> bool:
```

File: prowler/lib/scan_filters/scan_filters.py [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    """
+    Check if it is a tags-based
+
+    Returns True if it is filtered and False if it does not match the input filters
+    """
+    try:
+        if resource in str(audit_resources):
```
Comment (jfagoagas) [inline prowler/lib/scan_filters/scan_filters.py:11]:
```suggestion
        if resource in str(filtered_audit_resources):
```

File: prowler/providers/aws/lib/audit_info/audit_info.py [hunk 7]
```diff
@@ -21,6 +21,7 @@
         session_duration=None,
         external_id=None,
     ),
+    audit_resources=None,
```
Comment (jfagoagas) [inline prowler/providers/aws/lib/audit_info/audit_info.py:24]: I prefer to give this variable a more expressive name like `filtered_audit_resources` or `included_audit_resources`

File: tests/lib/scan_filters/scan_filters_test.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+            "arn:aws:s3:::test_bucket",
+        ]
+        assert is_resource_filtered(
+            "arn:aws:iam::123456789012:user/test_user", audit_resources
+        )
+        assert not is_resource_filtered(
+            "arn:aws:iam::123456789012:user/test1", audit_resources
+        )
+        assert is_resource_filtered("test_bucket", audit_resources)
+        assert is_resource_filtered("arn:aws:s3:::test_bucket", audit_resources)
```
Comment (jfagoagas) [inline tests/lib/scan_filters/scan_filters_test.py:17]: Please, split these behaviours into different tests.

File: tests/lib/scan_filters/scan_filters_test.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+            "arn:aws:iam::123456789012:user/test_user",
+            "arn:aws:s3:::test_bucket",
+        ]
+        assert is_resource_filtered(
+            "arn:aws:iam::123456789012:user/test_user", audit_resources
+        )
+        assert not is_resource_filtered(
+            "arn:aws:iam::123456789012:user/test1", audit_resources
+        )
+        assert is_resource_filtered("test_bucket", audit_resources)
```
Comment (jfagoagas) [inline tests/lib/scan_filters/scan_filters_test.py:16]: Why `test_bucket` without the ARN is included?

File: prowler/providers/common/audit_info.py [hunk 10]
```diff
@@ -236,6 +236,25 @@ def set_aws_audit_info(self, arguments) -> AWS_Audit_Info:
...
+            tagged_resources = []
+            for region in current_audit_info.audited_regions:
+                client = current_audit_info.audit_session.client(
+                    "resourcegroupstaggingapi", region_name=region
+                )
+                get_resources_paginator = client.get_paginator("get_resources")
+                for page in get_resources_paginator.paginate(TagFilters=scan_tags):
+                    for resource in page["ResourceTagMappingList"]:
+                        tagged_resources.append(resource["ResourceARN"])
+            current_audit_info.audit_resources = tagged_resources
```
Comment (jfagoagas) [inline prowler/providers/common/audit_info.py]: Please, include this code block as a separate function a include test for this function. Thanks!

**Review by jfagoagas — CHANGES_REQUESTED** [review]
Really good work here, but as an overall comment if we are including the new `audit_resources` in each service, hence in each test, we need to include new tests for each service to handle this behaviour. Including just a `audit_resources = None` is not enough.

File: prowler/lib/cli/parser.py (hunk 3 above)
Comment (MrCloudSec) [reply prowler/lib/cli/parser.py:354]: Done

File: prowler/providers/common/audit_info.py (hunk 10 above)
Comment (MrCloudSec) [reply prowler/providers/common/audit_info.py]: Done

File: prowler/providers/common/audit_info.py [hunk 11]
```diff
@@ -287,3 +292,26 @@ def set_provider_audit_info(provider: str, arguments: dict):
         sys.exit()
     else:
         return provider_audit_info
+
+
+def get_tagged_resources(input_scan_tags: list, current_audit_info: AWS_Audit_Info):
+    """
+    get_tagged_resources get the resources that have the tags input by the user and are going to be scanned
```
Comment (jfagoagas) [inline prowler/providers/common/audit_info.py]:
```suggestion
    get_tagged_resources returns a list of the resources that are going to be scanned based on the given input tags
    ```

File: prowler/providers/common/audit_info.py [hunk 12]
```diff
@@ -287,3 +292,26 @@ def set_provider_audit_info(provider: str, arguments: dict):
...
+        # Get Resources with scan_tags for all regions
+        for region in current_audit_info.audited_regions:
+            client = current_audit_info.audit_session.client(
+                "resourcegroupstaggingapi", region_name=region
+            )
+            get_resources_paginator = client.get_paginator("get_resources")
+            for page in get_resources_paginator.paginate(TagFilters=scan_tags):
+                for resource in page["ResourceTagMappingList"]:
+                    tagged_resources.append(resource["ResourceARN"])
+    return tagged_resources
```
Comment (jfagoagas) [inline prowler/providers/common/audit_info.py]: Include a `try/catch` block here and handle the possible errors. We need to include it in each function.

File: tests/lib/cli/parser_test.py [hunk 13]
```diff
@@ -795,6 +796,24 @@ def test_aws_parser_allowlist_long(self):
...
+    def test_aws_parser_scan_tags_long(self):
+        argument = "--scan-tags"
+        scan_tag1 = "Key=Value"
+        scan_tag2 = "Key2=Value2"
+        command = [prowler_command, argument, scan_tag1, scan_tag2]
+        parsed = self.parser.parse(command)
+        assert len(parsed.scan_tags) == 2
+        assert scan_tag1 in parsed.scan_tags
+        assert scan_tag2 in parsed.scan_tags
+
```
Comment (jfagoagas) [inline tests/lib/cli/parser_test.py:816]: 🔝

File: tests/providers/common/audit_info_test.py [hunk 14]
```diff
@@ -198,3 +210,40 @@ def test_set_audit_info_azure(self):
...
+        ]
+        client.create_tags(Resources=[image_id], Tags=[{"Key": "ami", "Value": "test"}])
+
+        mock_current_audit_info.audited_regions = ["eu-central-1"]
+        mock_current_audit_info.audit_session = boto3.session.Session()
+        assert len(get_tagged_resources(["ami=test"], mock_current_audit_info)) == 2
+        assert (
+            len(get_tagged_resources(["MY_TAG1=MY_VALUE1"], mock_current_audit_info))
+            == 1
+        )
```
Comment (jfagoagas) [inline tests/providers/common/audit_info_test.py]: Can you split this into two separate test? Thanks!

**Review by jfagoagas — CHANGES_REQUESTED** [review]

File: tests/providers/common/audit_info_test.py (hunk 14 above)
Comment (MrCloudSec) [reply tests/providers/common/audit_info_test.py]: I don't feel it is necessary, it is a same environment for different input tags. I have added more asserts though.

File: prowler/providers/common/audit_info.py (hunk 12 above)
Comment (MrCloudSec) [reply prowler/providers/common/audit_info.py]: Done

File: tests/providers/common/audit_info_test.py (hunk 14 above)
Comment (jfagoagas) [reply tests/providers/common/audit_info_test.py]: Cool

**Review by jfagoagas — APPROVED** [review]

**Pull request merged**